diskqueue.go

```diff
@@ -395,6 +436,195 @@ func (d *diskQueue) readOne() ([]byte, error) {
+	totalMessages, err := d.readNumOfMessages(curFileName)
+	if err != nil {
+		return err
+	}
+
+	// update depth with the remaining number of messages
+	d.depth -= totalMessages - d.readMessages
+
+	// we have not finished reading this file
+	if d.readFileNum == d.nextReadFileNum {
+		d.nextReadFileNum++
+		d.nextReadPos = 0
+	}
+
+	d.moveToNextReadFile()
+
+	return nil
+}
+
+// walk through all of the files in the DiskQueue directory
+func (d *diskQueue) walkDiskQueueDir(fn func(os.FileInfo) error) error {
+	fileInfos, err := ioutil.ReadDir(d.dataPath)
+
+	if err != nil {
+		return err
+	}
+
+	for _, fileInfo := range fileInfos {
+		// only go through files and skip directories
+		if !fileInfo.IsDir() {
+			err = fn(fileInfo)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]os.FileInfo, error) {
+	var badFileInfos []os.FileInfo
+
+	getAllBadFileInfo := func(fileInfo os.FileInfo) error {
+		// only accept "bad" files created by this DiskQueue object
+		if badFileNameRegexp.MatchString(fileInfo.Name()) {
+			badFileInfos = append(badFileInfos, fileInfo)
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateTotalDiskSpaceUsed() {
+	d.totalDiskSpaceUsed = maxMetaDataFileSize
+
+	updateTotalDiskSpaceUsed := func(fileInfo os.FileInfo) error {
+		// only accept files created by this DiskQueue object
+		if fileNameRegexp.MatchString(fileInfo.Name()) || badFileNameRegexp.MatchString(fileInfo.Name()) {
+			d.totalDiskSpaceUsed += fileInfo.Size()
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(updateTotalDiskSpaceUsed)
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to update write bytes - %s", d.name, err)
+	}
+}
+
+func (d *diskQueue) freeDiskSpace(expectedBytesIncrease int64) error {
+	var err error
+	var badFileInfos []os.FileInfo
+
+	badFileInfos, err = d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
+		if len(badFileInfos) > 0 {
+			// check if a .bad file exists. If it does, delete that first
+			err = d.removeBadFile(badFileInfos[0])
+			if err != nil {
+				return err
+			}
+			badFileInfos = badFileInfos[1:]
+		} else {
+			// delete the read file (make space)
+			readFileToDeleteNum := d.readFileNum
+			err = d.removeReadFile()
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove file(%s) - %s", d.name, d.fileName(readFileToDeleteNum), err)
+				d.handleReadError()
+				return err
+			} else {
+				d.logf(INFO, "DISKQUEUE(%s) removed file(%s) to free up disk space", d.name, d.fileName(readFileToDeleteNum))
+			}
+			d.updateTotalDiskSpaceUsed()
+		}
+	}
+
```

Comment:
To avoid possible infinite loop (due to unsynced variables or other cases), I would prefer looping each possible files and break if there is enough space.
```go
	if expectedBytesIncrease > d.maxBytesDiskSpace {
		return fmt.Errorf("could not make space for expectedBytesIncrease = %d, with maxBytesDiskSpace = %d ", expectedBytesIncrease, d.maxBytesDiskSpace)
	}

	// keep freeing up disk space until we have enough space to write this message
	for _, badFileInfo := range badFileInfos {
		if d.totalDiskSpaceUsed+expectedBytesIncrease <= d.maxBytesDiskSpace {
			break
		}
		d.deleteBadFile(badFileInfo)
	}
	for d.readFileNum <= d.writeFileNum {
		if d.totalDiskSpaceUsed+expectedBytesIncrease <= d.maxBytesDiskSpace {
			break
		}
		// delete the read file (make space)
		readFileToDeleteNum := d.readFileNum
		err = d.removeReadFile()
		if err != nil {
			d.logf(ERROR, "DISKQUEUE(%s) failed to remove file(%s) - %s", d.name, d.fileName(readFileToDeleteNum), err)
			d.handleReadError()
			return err
		} else {
			d.logf(INFO, "DISKQUEUE(%s) removed file(%s) to free up disk space", d.name, d.fileName(readFileToDeleteNum))
		}
		d.updateTotalDiskSpaceUsed()
	}

	if d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
		return fmt.Errorf("could not make space for totalDiskSpaceUsed = %d, expectedBytesIncrease = %d, with maxBytesDiskSpace = %d ", d.totalDiskSpaceUsed, expectedBytesIncrease, d.maxBytesDiskSpace)
	}
```

Reply:
Good idea!